### PR TITLE
Clippy lints + clippy CI

### PR DESCRIPTION
**Pull request by joshfried-aws**
*Issue #, if available:*
#208 

*Description of changes:*
As part of our goal to improve code quality, we are addressing all clippy lints in our project + adding a clippy check to our CI

---
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: guard/src/commands/validate/common.rs [hunk 1]
```diff
@@ -334,50 +331,6 @@ pub(super) fn extract_name_info_from_record<'record, 'value>(
     })
 }
 
-pub(crate) fn extract_event_records<'value>(
```
Comment (akshayrane) [inline guard/src/commands/validate/common.rs:337]: Unused method?

File: guard/src/commands/validate/common.rs [hunk 2]
```diff
@@ -661,13 +606,12 @@ where
                                     } else {
                                         "was int"
                                     },
-                                IsFloat =>
+                                _ =>
```
Comment (akshayrane) [inline guard/src/commands/validate/common.rs:609]: Is this change expected?

Looks like `_` is the default arm, but we have messages that are related to `cmp` being a float?

File: guard/src/commands/validate/console_reporter.rs [hunk 3]
```diff
@@ -263,7 +262,7 @@ fn pprint_failed_sub_tree(
                                 to=to_result.map_or("NULL".to_string(), |t| format!("{}", t)),
                                 op_msg=match cmp {
                                     CmpOperator::Eq => if *not { "equal to" } else { "not equal to" },
-                                    CmpOperator::Le => if *not { "less than equal to" } else { "less than equal to" },
+                                    CmpOperator::Le => if *not { "less than equal to" } else { "not less than equal to" },
```
Comment (akshayrane) [inline guard/src/commands/validate/console_reporter.rs:265]: Minor bugfix! Good catch!

File: guard/src/commands/validate/tf.rs [hunk 4]
```diff
@@ -55,22 +51,20 @@ impl<'reporter> Reporter for TfAware<'reporter> {
...
-            Ok(match output_type {
+            match output_type {
                 OutputFormatType::YAML => serde_yaml::to_writer(write, &failure_report)?,
                 OutputFormatType::JSON => serde_json::to_writer_pretty(write, &failure_report)?,
                 OutputFormatType::SingleLineSummary => {
                     single_line(write, data_file, rules_file, data, root, failure_report)?
                 }
-            })
+            };
+            Ok(())
```
Comment (akshayrane) [inline guard/src/commands/validate/tf.rs:67]: Same question as before: are we changing the return type?

File: guard/src/commands/validate/tf.rs [hunk 5]
```diff
@@ -276,14 +259,14 @@ fn single_line(
                         re: &UnaryComparison<'_>,
                         prefix: &str,
                     ) -> crate::rules::Result<usize> {
-                        let width = "PropertyPath".len() + 4;
+                        let _width = "PropertyPath".len() + 4;
```
Comment (akshayrane) [inline guard/src/commands/validate/tf.rs:262]: Okay, so this variable is being used and still has a prefix? 

Is [this](https://stackoverflow.com/a/48361729) the correct explanation?

File: guard/src/rules/eval.rs [hunk 6]
```diff
@@ -59,10 +59,14 @@ is_type_fn!(is_string_operation, PathAwareValue::String(_));
 is_type_fn!(is_list_operation, PathAwareValue::List(_));
 is_type_fn!(is_struct_operation, PathAwareValue::Map(_));
 is_type_fn!(is_int_operation, PathAwareValue::Int(_));
+#[cfg(test)]
```
Comment (akshayrane) [inline guard/src/rules/eval.rs:62]: If these are tests, should we move them to `eval_tests.rs`?

File: guard/src/rules/values_tests.rs [hunk 7]
```diff
@@ -177,7 +177,6 @@ fn test_query_on_value() -> Result<()> {
         if let PathAwareValue::Map(_index) = each {
             continue;
         }
-        assert!(false);
```
Comment (akshayrane) [inline guard/src/rules/values_tests.rs:180]: I think this might be an exit for a negative test? Like if code reaches here we want to fail the test right?

File: guard/src/rules/path_value.rs [hunk 8]
```diff
@@ -217,57 +191,6 @@ pub(crate) enum PathAwareValue {
     RangeChar((Path, RangeType<char>)),
 }
 
-impl PathAwareValue {
```
Comment (akshayrane) [inline guard/src/rules/path_value.rs:220]: unused snippet?

File: guard/src/rules/path_value.rs [hunk 9]
```diff
@@ -122,41 +124,22 @@ impl Path {
...
-
-    pub(crate) fn extend_string(&self, part: &String) -> Path {
-        self.extend_str(part.as_str())
+    //
+    // pub(crate) fn extend_str_with_location(&self, part: &str, loc: Location) -> Path {
+    //     let mut copy = self.0.clone();
+    //     copy.push('/');
+    //     copy.push_str(part);
+    //     Path(copy, loc)
+    // }
```
Comment (akshayrane) [inline guard/src/rules/path_value.rs:133]: Why is this chunk commented out?

File: guard/src/rules/functions/strings.rs [hunk 10]
```diff
@@ -1,194 +0,0 @@
-use crate::rules::path_value::{Path, PathAwareValue};
```
Comment (akshayrane) [inline guard/src/rules/functions/strings.rs:1]: Why are these removed?

File: guard/src/rules/functions/collections.rs [hunk 11]
```diff
@@ -1,12 +0,0 @@
-use crate::rules::QueryResult;
```
Comment (akshayrane) [inline guard/src/rules/functions/collections.rs:1]: Why are we removing these?

File: guard/src/rules/eval_context_tests.rs [hunk 12]
```diff
@@ -28,11 +28,7 @@ fn extraction_test() -> Result<()> {
             .rules
             .get("aws_route53_recordset")
             .map(|s| s.get(0))
-            .map(|s| match s {
-                Some(r) => Some(*r),
-                None => None,
-            })
-            .flatten(),
+            .and_then(|s| s.copied()),
```
Comment (akshayrane) [inline guard/src/rules/eval_context_tests.rs:31]: Wait, is there a logic change here?

File: guard/src/rules/eval/operators.rs [hunk 13]
```diff
@@ -115,18 +106,18 @@ pub(crate) struct NotComparable<'r> {
...
-) -> std::result::Result<&'value PathAwareValue, R>
-where
-    E: Fn(UnResolved<'value>) -> R,
-{
-    match qr {
-        QueryResult::Resolved(r) | QueryResult::Literal(r) => Ok(*r),
-        QueryResult::UnResolved(ur) => Err(err(ur.clone())),
-    }
-}
+// pub(super) fn resolved<'value, E, R>(
```
Comment (akshayrane) [inline guard/src/rules/eval/operators.rs:109]: Why is this chunk commented out?

File: guard/src/commands/helper.rs [hunk 14]
```diff
@@ -17,20 +17,21 @@ pub struct ValidateInput<'a> {
...
 }
 
+#[allow(dead_code)]
 pub fn validate_and_return_json(
     data: ValidateInput,
     rules: ValidateInput,
     verbose: bool,
 ) -> Result<String> {
-    let input_data = match serde_json::from_str::<serde_json::Value>(&data.content) {
+    let input_data = match serde_json::from_str::<serde_json::Value>(data.content) {
```
[the remark this answers is not in the document]
Comment (joshfried-aws) [reply guard/src/commands/helper.rs:26]: No it doesnt...Since the content field of ValidateInput is already a `&str` the reference infront of `&data.content` is immediately being dereferenced by the compiler

File: guard/src/commands/validate/cfn.rs [hunk 15]
```diff
@@ -465,7 +432,7 @@ fn handle_resource_aggr<'record, 'value: 'record>(
     root: &'value Node<'_>,
     name: String,
     by_resources: &mut HashMap<String, LocalResourceAggr<'record, 'value>>,
-    value: &Vec<Rc<crate::commands::validate::common::Node<'record, 'value>>>,
+    value: &[Rc<crate::commands::validate::common::Node<'record, 'value>>],
```
[the remark this answers is not in the document]
Comment (joshfried-aws) [reply guard/src/commands/validate/cfn.rs:435]: Technically yes, but it doesnt require any change other than function signature. https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

File: guard/src/commands/validate/cfn_reporter.rs [hunk 16]
```diff
@@ -181,7 +172,7 @@ impl super::common::GenericReporter for SingleLineReporter {
         by_resource_name: HashMap<String, Vec<NameInfo<'_>>>,
         passed: HashSet<String>,
         skipped: HashSet<String>,
-        longest_rule_len: usize,
+        _longest_rule_len: usize,
```
[the remark this answers is not in the document]
Comment (joshfried-aws) [reply guard/src/commands/validate/cfn_reporter.rs:175]: confirmed

File: guard/src/rules/functions/strings.rs (hunk 10 above)
Comment (joshfried-aws) [reply guard/src/rules/functions/strings.rs:1]: Compiler was complaining about unused code, and i commented it out and ran the test suite + did manual testing to make sure it wasnt actually being used...Forgot to delete it. Good catch

File: guard/src/rules/path_value.rs (hunk 9 above)
Comment (joshfried-aws) [reply guard/src/rules/path_value.rs:133]: Compiler was complaining about unused code, and i commented it out and ran the test suite + did manual testing to make sure it wasnt actually being used...Forgot to delete it. Good catch

File: guard/src/rules/functions/collections.rs (hunk 11 above)
Comment (joshfried-aws) [reply guard/src/rules/functions/collections.rs:1]: Wasnt being used anywhere

File: guard/src/rules/eval/operators.rs (hunk 13 above)
Comment (joshfried-aws) [reply guard/src/rules/eval/operators.rs:109]: same as the previous commented out chunk

File: guard/src/commands/validate/common.rs (hunk 1 above)
Comment (joshfried-aws) [reply guard/src/commands/validate/common.rs:337]: yes

File: guard/src/commands/validate/tf.rs (hunk 4 above)
Comment (joshfried-aws) [reply guard/src/commands/validate/tf.rs:67]: same answer., we're not changing the return type. But this is exactly why its best practice to return unit types this way

File: guard/src/commands/validate/tf.rs (hunk 5 above)
Comment (joshfried-aws) [reply guard/src/commands/validate/tf.rs:262]: It's not being used, I should have deleted it...For some reason it was defined there and then again [7 lines later](https://github.com/aws-cloudformation/cloudformation-guard/blob/main/guard/src/commands/validate/tf.rs#L286)

File: guard/src/rules/eval_context_tests.rs (hunk 12 above)
Comment (joshfried-aws) [reply guard/src/rules/eval_context_tests.rs:31]: nope, see: https://doc.rust-lang.org/rust-by-example/error/option_unwrap/and_then.html

File: guard/src/rules/values_tests.rs (hunk 7 above)
Comment (joshfried-aws) [reply guard/src/rules/values_tests.rs:180]: Possibly, but then wouldnt it be best to do something like

```
assert!(matches!(PathAwareValue::Map(_), each)); 
```

File: guard/src/commands/validate/common.rs (hunk 2 above)
Comment (joshfried-aws) [reply guard/src/commands/validate/common.rs:609]: CmpOperator doesnt have an IsFloat variant

File: guard/src/commands/validate/cfn.rs [hunk 17]
```diff
@@ -98,7 +94,9 @@ impl<'reporter> Reporter for CfnAware<'reporter> {
                     data,
                     failure_report,
                 )?,
-            })
+            };
+
+            Ok(())
```
[the remark this answers is not in the document]
Comment (brayalad) [reply guard/src/commands/validate/cfn.rs:99]: Kotlin has 'Unit' too

File: .github/workflows/pr.yml [hunk 18]
```diff
@@ -2,9 +2,9 @@ name: Rust
 
 on:
   push:
-    branches: [ main, development, rogue_one]
+    branches: [ main, development, rogue_one ]
```
Comment (andrewjeminchoi) [inline .github/workflows/pr.yml:5]: 👍

File: guard/src/rules/path_value.rs [hunk 19]
```diff
@@ -122,41 +124,22 @@ impl Path {
...
-    pub(crate) fn extend_str_with_location(&self, part: &str, loc: Location) -> Path {
-        let mut copy = self.0.clone();
-        copy.push('/');
-        copy.push_str(part);
-        Path(copy, loc)
-    }
-
-    pub(crate) fn extend_string(&self, part: &String) -> Path {
-        self.extend_str(part.as_str())
+    //
```
Comment (andrewjeminchoi) [inline guard/src/rules/path_value.rs:127]: Please consider deletion

**Comment by joshfried-aws** [comment]
rejecting this in favour of #371
